### PR TITLE
[llvm] Recognize GEP

### PR DESCRIPTION
**Pull request by EgorBo**
```csharp
static byte Foo1(byte* array)
{
    return array[42];
}

static byte Foo2(byte* array, int i)
{
    return array[i];
}
```

### Old IR (after opt)
```llvm
; Function Attrs: noinline norecurse nounwind readonly uwtable
define hidden monocc i8 @Program_Foo1_byte_(i8* nocapture readonly %arg_myArray) #4 {
BB0:
  %0 = getelementptr i8, i8* %arg_myArray, i64 42 ;; It's OK!
  %1 = load i8, i8* %0, align 1
  ret i8 %1
}

; Function Attrs: noinline norecurse nounwind readonly uwtable
define hidden monocc i8 @Program_Foo2_byte__int(i8* %arg_myArray, i32 %arg_i) #4 {
BB0:
  %t21 = sext i32 %arg_i to i64
  %0 = ptrtoint i8* %arg_myArray to i64 ;; should be GEP
  %t22 = add i64 %t21, %0
  %1 = inttoptr i64 %t22 to i8*
  %2 = load i8, i8* %1, align 1
  ret i8 %2
}
```

### New IR (after opt)
```llvm
; Function Attrs: noinline norecurse nounwind readonly uwtable
define hidden monocc i8 @Program_Foo1_byte_(i8* nocapture readonly %arg_myArray) #4 {
BB0:
  %0 = getelementptr i8, i8* %arg_myArray, i64 42
  %1 = load i8, i8* %0, align 1
  ret i8 %1
}

; Function Attrs: noinline norecurse nounwind readonly uwtable
define hidden monocc i8 @Program_Foo2_byte__int(i8* nocapture readonly %arg_myArray, i32 %arg_i) #4 {
BB0:
  %t21 = sext i32 %arg_i to i64
  %0 = getelementptr i8, i8* %arg_myArray, i64 %t21
  %1 = load i8, i8* %0, align 1
  ret i8 %1
}
```

So instead of `ptrtoint+add+inttoptr` we can (should) emit `getelementptr` which is technically the same but more opt-friendly, e.g. loop-vectorization won't work without it.
We used to emit it only when offset is a const (see Foo1).

Now LLVM finally recognizes `memset` in the following C# code!:
```csharp
static void Foo3(byte* array, int len, byte val)
{
    for (int i = 0; i < len; i++)
        array[i] = val;
}
```
### LLVM IR:
```llvm
; Function Attrs: noinline norecurse nounwind uwtable
define hidden monocc void @Program_Foo3_byte__int_byte(i8* nocapture %arg_myArray, i32 %arg_len, i32 %arg_val) #5 {
BB0:
  %0 = icmp sgt i32 %arg_len, 0
  br i1 %0, label %BB5.lr.ph, label %BB1

BB5.lr.ph:                                        ; preds = %BB0
  %1 = trunc i32 %arg_val to i8
  %2 = zext i32 %arg_len to i64
  call void @llvm.memset.p0i8.i64(i8* %arg_myArray, i8 %1, i64 %2, i32 1, i1 false)
  br label %BB1

BB1:                                              ; preds = %BB5.lr.ph, %BB0
  ret void
}
```

### CODE REVIEW

File: mono/mini/llvm-jit.cpp [hunk 1]
```diff
@@ -297,12 +297,13 @@ class MonoLLVMJIT {
...
 		initializeTarget(registry);
+		initializeLoopIdiomRecognizeLegacyPassPass(registry);
 		linkCoreCLRGC(); // Mono uses built-in "coreclr" GCStrategy
 
 		// FIXME: find optimal mono specific order of passes
 		// see https://llvm.org/docs/Frontend/PerformanceTips.html#pass-ordering
 		// the following order is based on a stripped version of "OPT -O2"
-		const char *default_opts = " -simplifycfg -sroa -lower-expect -instcombine -licm -simplifycfg -lcssa -indvars -loop-deletion -gvn -memcpyopt -sccp -bdce -instcombine -dse -simplifycfg -enable-implicit-null-checks";
+		const char *default_opts = " -simplifycfg -sroa -lower-expect -instcombine -loop-rotate -licm -simplifycfg -lcssa -loop-idiom -indvars -loop-deletion -gvn -memcpyopt -sccp -bdce -instcombine -dse -simplifycfg -enable-implicit-null-checks";
 		const char *opts = g_getenv ("MONO_LLVM_OPT");
```
Comment (EgorBo) [inline mono/mini/llvm-jit.cpp]: `-loop-idiom` is the pass that recognizes memset in a loop (also needs `-loop-rotate`). So now both AOT and JIT can do it.

Comment (EgorBo) [reply mono/mini/llvm-jit.cpp]: image

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
Local tests: https://gist.github.com/EgorBo/cf763a7079ad218c574e5719aae75fd3

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by vargaz** [comment]
@monojenkins squash

**Pull request merged**